execution-scheduler/execution_scheduler/main.py

```diff
@@ -0,0 +1,116 @@
+from manager_rest import manager_exceptions
+from manager_rest.storage import models, get_storage_manager
+from manager_rest.utils import set_current_tenant
+from manager_rest.flask_utils import setup_flask_app
+from manager_rest.resource_manager import get_resource_manager
+from manager_rest.storage.storage_utils import (try_acquire_lock_on_table,
+                                                unlock_table)
+
+
+logger = logging.getLogger(__name__)
+BROKER_PORT_SSL = 5671
+BROKER_PORT_NO_SSL = 5672
+SCHEDULER_LOCK_BASE = 10000
+
+DEFAULT_LOG_PATH = '/var/log/cloudify/execution_scheduler/scheduler.log'
+CONFIG_PATH = '/opt/manager/cloudify-rest.conf'
+REST_SECURITY_PATH = '/opt/manager/rest-security.conf'
+
+
+def check_schedules():
+    logging.info('Checking schedules...')
+    sm = get_storage_manager()
+    schedules = sm.list(
+        models.ExecutionSchedule,
+        filters={'next_occurrence': lambda x: x < datetime.utcnow()}
+    )
+    for schedule in schedules:
+        next_occurrence = schedule.compute_next_occurrence()
+        old_next_occurrence = schedule.next_occurrence
+
+        logging.info('Schedule: %s next in %s; old next was %s',
+                     schedule.id, next_occurrence, old_next_occurrence)
+
+        lock_num = SCHEDULER_LOCK_BASE + schedule._storage_id
+        with scheduler_lock(lock_num) as locked:
+            if not locked:
+                logger.info('Another manager currently runs this schedule')
+                continue
+            logger.debug('Acquired lock for schedule %s in DB', schedule.id)
+            if should_run(schedule):
+                execution = execute_workflow(schedule)
+                schedule.latest_execution = execution
+            schedule.next_occurrence = next_occurrence
+            sm.update(schedule)
+
+
+@contextmanager
+def scheduler_lock(lock_number):
+    locked = try_acquire_lock_on_table(lock_number)
+    try:
+        yield locked
+    finally:
+        if locked:
+            unlock_table(lock_number)
+
+
+def should_run(schedule):
+    # TODO :: also, don't run if the latest execution is not in an END_STATE
+    slip = timedelta(minutes=schedule.slip, seconds=59)
+    next_occurrence = dateutil.parser.parse(schedule.next_occurrence,
+                                            ignoretz=True)
+    return datetime.utcnow() - next_occurrence < slip
+
+
+def execute_workflow(schedule):
+    rm = get_resource_manager()
+    set_current_tenant(schedule.tenant)
+    logging.info('Running: %s', schedule)
+
+    # always queue an execution which couldn't start
+    execution_arguments = schedule.execution_arguments or {}
+    execution_arguments['queue'] = True
+
+    execution = rm.execute_workflow(
+        deployment_id=schedule.deployment.id,
+        workflow_id=schedule.workflow_id,
+        parameters=schedule.parameters,
+        execution_creator=schedule.creator,
+        **schedule.execution_arguments
+    )
+    set_current_tenant(None)
```

Comment:
it would be nicer to have this in a `finally`, in case something throws, so that we dont leave a tenant set; probably use a contextmanager